src/controller/index.js

```diff
@@ -116,15 +137,186 @@ class Controller extends ENIP {
         this.state.controller.path = PORT.build(BACKPLANE, SLOT);
 
         const sessid = await super.connect(IP_ADDR);
-
         if (!sessid) throw new Error("Failed to Register Session with Controller");
+        
+        this._initializeControllerEventHandlers(); // Connect sendRRData Event
 
-        this._initializeControllerEventHandlers();
+        if (this.state.connectedMessaging === true) {
+            const connid = await this.forwardOpen();
+            if(!connid) throw new Error("Failed to Forward Open with Controller");
+        }
 
         // Fetch Controller Properties and Wall Clock
         await this.readControllerProps();
     }
 
+    /**
+     * Disconnects the PLC instance gracefully by issuing forwardClose, UnregisterSession
+     * and then destroying the socket
+     * and Returns a Promise indicating a success or failure or the disconnection
+     *
+     * @memberof Controller
+     * @returns {Promise}
+     */
+    async disconnect() {
+        if (super.established_conn === true) {
+            const closeid = await this.forwardClose();
+            if(!closeid) throw new Error("Failed to Forward Open with Controller");
+        }
+
+        super.destroy();
+
+        this._removeControllerEventHandlers();
+        return "disconnected";
+    }
+
+    /**
+     * Writes a forwardOpen Request and retrieves the connection ID used for
+     * connected messages.
+     * @memberof Controller
+     * @returns {Promise}
+     */
+    async forwardOpen() {
```

Comment:
Should the `forwardOpen` and `forwardClose` methods be moved up the inheritance chain into the standard `ENIP` class?  Or is this particular implementation specific to Rockwell PLCs?